Switch docker image to alpine

avLightless opened the pull request:
**What's changed**

- Added build step
- Modified go build flags
  - Added ldflags to strip debugging information
  - Changed `goos` to linux
  - Disabled `cgo` at compile time
- Changed final image to `alpine`
- Added non privileged user

**Result:** 

- Docker image size went from `960MB` to `13.9MB`
- App no longer runs as root
- The image only contains the basic alpine components and the compiled static binary of cryptgo.

avLightless committed:
Update Dockerfile

Gituser143 commented:
Hi @avLightless! Thank you for your PR! I'm not sure what the `"-s -w"` build flags passed to `--ldflags` do. Could you point me towards any resource so that I can understand what they do?

Additionally, `cryptgo` requires mounting of the user's home directory to store some metadata (`~/.cryptgo-data.json`), since now the container no longer runs with the user as root, the docker command must be appropriately changed to mount the home directory to the appropriate place (`/home/appuser` instead of `/root`).

avLightless commented:
> Hi @avLightless! Thank you for your PR! I'm not sure what the `"-s -w"` build flags passed to `--ldflags` do. Could you point me towards any resource so that I can understand what they do?
> 
> Additionally, `cryptgo` requires mounting of the user's home directory to store some metadata (`~/.cryptgo-data.json`), since now the container no longer runs with the user as root, the docker command must be appropriately changed to mount the home directory to the appropriate place (`/home/appuser` instead of `/root`).

Ofcourse. If you run `go tool link` you can see `-s	disable symbol table` and `-w	disable DWARF generation` at the bottom.
Here are some additional links where this is mentioned:

- https://medium.com/a-journey-with-go/go-how-to-take-advantage-of-the-symbols-table-360dd52269e5
- https://lukeeckley.com/post/useful-go-build-flags/
- https://stackoverflow.com/questions/22267189/what-does-the-w-flag-mean-when-passed-in-via-the-ldflags-option-to-the-go-comman

As stated in some of the above links, this should have no effect on the application besides being unable to debug it.

As for the mounting, you are completely correct. I also recommend using `--rm` as without it, the container lingers around in the system after closing it, and would require a `docker start / docker attach` instead of `docker run`. I will update the readme in my PR soon.

We also made a gitlab repo - for personal use -  which builds this project once a week as a docker image directly from your github repo (just with a modified Dockerfile), and puts the image in the registry. Take a look if you want to. I don't know if github can do the same, but it would be nice if it could build the image every time you push an update :)
https://gitlab.com/avLightless/cryptgo-builder/container_registry/2200931

avLightless committed:
Updated docker mount directory

New mount directory accomodates the unprivileged user of the alpine based image.

Gituser143 commented:
> > Hi @avLightless! Thank you for your PR! I'm not sure what the `"-s -w"` build flags passed to `--ldflags` do. Could you point me towards any resource so that I can understand what they do?
> > Additionally, `cryptgo` requires mounting of the user's home directory to store some metadata (`~/.cryptgo-data.json`), since now the container no longer runs with the user as root, the docker command must be appropriately changed to mount the home directory to the appropriate place (`/home/appuser` instead of `/root`).
> 
> Ofcourse. If you run `go tool link` you can see `-s disable symbol table` and `-w disable DWARF generation` at the bottom.
> Here are some additional links where this is mentioned:
> 
>     * https://medium.com/a-journey-with-go/go-how-to-take-advantage-of-the-symbols-table-360dd52269e5
> 
>     * https://lukeeckley.com/post/useful-go-build-flags/
> 
>     * https://stackoverflow.com/questions/22267189/what-does-the-w-flag-mean-when-passed-in-via-the-ldflags-option-to-the-go-comman
> 
> 
> As stated in some of the above links, this should have no effect on the application besides being unable to debug it.

Thanks, this helps!

> As for the mounting, you are completely correct. I also recommend using `--rm` as without it, the container lingers around in the system after closing it, and would require a `docker start / docker attach` instead of `docker run`. I will update the readme in my PR soon.

Yep that makes perfect sense, I don't know why I hadn't thought of that previously.

> We also made a gitlab repo - for personal use - which builds this project once a week as a docker image directly from your github repo (just with a modified Dockerfile), and puts the image in the registry. Take a look if you want to. I don't know if github can do the same, but it would be nice if it could build the image every time you push an update :)
> https://gitlab.com/avLightless/cryptgo-builder/container_registry/2200931

GitHub does have functionality to build container images, it has it's own container registry too (ghcr.io). There are github actions that can be defined to build and publish an image either on push or release. I hadn't gotten to set that up yet because I felt the project wasn't mature enough, but it does make sense to set it up now. If you'd like to, you could do that through another PR and use this just to clean up the Dockerfile maybe? 

Here's a [link](https://github.com/pesos/grofer/blob/main/.github/workflows/docker-publish.yml) to an existing action which I've used that publishes an image.

Edit: After a bit of reading I feel pushing to dockerhub might be a better idea as github's container registry has a limit on storage. This can be done through actions as well.

Gituser143 commented:
@avLightless Changes look good to me! Unless you have anything else to add. I'll merge after your confirmation.

avLightless commented:
Feel free to merge @Gituser143. I will look into the ghcr sometime.
